File: 160.java
//160. Intersection of Two Linked Lists

public class Solution {

    public ListNode getIntersectionNodeTwoPointer(ListNode headA, ListNode headB) {
        ListNode a = headA, b = headB;
        // int i = 0;
        while (a != b) {
            // System.out.println((++i) + " " + (a == null ? "-" : a.val) + " " + (b == null
            // ? "-" : b.val));
            a = a == null ? headA : a.next;
            b = b == null ? headB : b.next;
        }
        return a;
    }

    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        Set<ListNode> setB = new HashSet<>();
        while (headB != null) {
            setB.add(headB);
            headB = headB.next;
        }
        while (headA != null) {
            if (setB.contains(headA))
                return headA;
            headA = headA.next;
        }
        return null;
    }
}

File: 160.js
//approach would be to just set pointers in respective linked list
//if one pointer comes to an node end, just set that pointer to the head of the other linked list
//continue that in a while loop until the node from both linked list are not equal.
var getIntersectionNode = function(headA, headB) {
    //check if any linked list is null, if they are then just return null;
    if(headA === null || headB === null) return null;
    
    let aPtr = headA;
    let bPtr = headB;
    
    while(aPtr != bPtr){
        //check if the aPtr is null, if null, reset the aPtr to point to bPtr;
        //else keep moving
        if(aPtr === null) aPtr = headB;
        else aPtr = aPtr.next;
        
        //check if the bPtr is null, if null, reset the aPtr to point to bPtr;
        //else keep moving
        if(bPtr === null) bPtr = headA;
        else bPtr = bPtr.next;
    }
    //returning a node that contains the start of the intersection 
    //OR
    //they both hit null, so that's what will return if the loop cant detect the intersection
    return aPtr; 
};

